### PR TITLE
Include Tight decoding optimizations from TurboVNC

### PR DESCRIPTION
**Pull request by dcommander**
- As with the encoder, the decoder now uses the TurboJPEG wrapper, which
  allows it to decode JPEG images directly into the framebuffer.  This
  eliminates a buffer copy (CopyRectangle()) as well as the expensive
  RGB pixel conversion in DecompressJpegRectBPP().  The TurboJPEG
  wrapper performs RGB pixel conversion more optimally, and only when
  necessary (it uses the libjpeg-turbo colorspace extensions when
  available, in order to avoid RGB conversion.)
- The other Tight subencoding types are also now decoded directly into
  the framebuffer, which eliminates buffer copies.
- The Tight decoder now supports the rfbTightNoZlib extension, which
  allows the server to bypass zlib compression when Compression Level 0
  is selected.  The encoder already supports this extension.  Passing
  the data stream through zlib when Compression Level 0 is selected
  needlessly wastes CPU time, since all zlib is doing is copying the
  data internally into its own structures.

### CODE REVIEW

**Comment by bk138** [comment]
Thanks very much for that! Can you pls have a look at the AppVeyor build?

**Comment by dcommander** [comment]
Fixed and re-pushed.  For some reason, AppVeyor did not pick up the altered commit, so it is still failing.  I don't know how to force it to rebuild the working commit (https://github.com/LibVNC/libvncserver/pull/203/commits/f9798b4309c31547e49cceaade43e46b6e71bd85).

**Comment by dcommander** [comment]
The CI issue seems to have resolved itself.

**Comment by bk138** [comment]
Kinda ;-) I triggered a new build.

**Comment by dcommander** [comment]
Any word on when these enhancements will be integrated?

**Comment by bk138** [comment]
aww i totally forgot about this one, sry. will try to review this month, but there's lotsa $$$-job work in the way :-/

File: rfb/rfbclient.h [hunk 1]
```diff
@@ -296,11 +298,7 @@ typedef struct _rfbClient {
 
 #ifdef LIBVNCSERVER_HAVE_LIBJPEG
 	/** JPEG decoder state. */
-	rfbBool jpegError;
```
Comment (bk138) [inline rfb/rfbclient.h]: @dcommander to not break ABI, those should be left in (but clearly marked obsolete and due for removal at some version bump)

File: rfb/rfbclient.h [hunk 2]
```diff
@@ -296,11 +298,7 @@ typedef struct _rfbClient {
...
-
-	struct jpeg_source_mgr* jpegSrcManager;
-	void* jpegBufferPtr;
-	size_t jpegBufferLen;
+	tjhandle tjhnd;
```
Comment (bk138) [inline rfb/rfbclient.h]: this should go to the end of the struct to keep ABI compatibility

File: CMakeLists.txt [hunk 3]
```diff
@@ -317,6 +317,13 @@ set(LIBVNCCLIENT_SOURCES
     ${COMMON_DIR}/minilzo.c
 )
 
+if(JPEG_FOUND)
```
Comment (bk138) [inline CMakeLists.txt]: As the encoder (and now the decoder as well) only works with libjpeg-turbo (am I right on this point?) the libjpeg differentiation in https://github.com/dcommander/libvncserver/blob/f9798b4309c31547e49cceaade43e46b6e71bd85/CMakeLists.txt#L59 can probably go away?

Comment (dcommander) [reply CMakeLists.txt]: No, they should both work with libjpeg as well.

**Comment by dcommander** [comment]
Re-pushed with the requested changes

File: rfb/rfbclient.h [hunk 4]
```diff
@@ -295,13 +297,16 @@ typedef struct _rfbClient {
...
-	/** JPEG decoder state. */
+	/** JPEG decoder state (obsolete-- do not use). */
 	rfbBool jpegError;
 
 	struct jpeg_source_mgr* jpegSrcManager;
 	void* jpegBufferPtr;
 	size_t jpegBufferLen;
 
+	/** JPEG decoder state. */
+	tjhandle tjhnd;
```
Comment (bk138) [inline rfb/rfbclient.h]: i meant: to the end of the struct. the way it is now, other struct members after this one are offset by sizeof(tjhandle).

**Comment by dcommander** [comment]
Re-pushed with the requested changes.

**Pull request merged**